exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -313,113 +322,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 			return err
 		}
 		wIndex := prwe.wWALIndex.Add(1)
+		prwe.log.Debug("write", zap.Uint64("index", wIndex))
 		batch.Write(wIndex, protoBlob)
 	}
 
+	// notify possibly waiting tailing routine of write
+	select {
+	case prwe.rNotify <- struct{}{}:
+		// tailing routine is actively waiting for write
+	default:
+		// no receiver, ignore
+	}
+
 	return prwe.wal.WriteBatch(batch)
 }
 
-func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq *prompb.WriteRequest, err error) {
-	prwe.mu.Lock()
-	defer prwe.mu.Unlock()
-
-	var protoBlob []byte
-	for i := 0; i < 12; i++ {
-		// Firstly check if we've been terminated, then exit if so.
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		case <-prwe.stopChan:
-			return nil, fmt.Errorf("attempt to read from WAL after stopped")
-		default:
-		}
-
-		if index <= 0 {
-			index = 1
-		}
-
-		if prwe.wal == nil {
-			return nil, fmt.Errorf("attempt to read from closed WAL")
-		}
+// read repeatedly attempts to fetch a *prompb.WriteRequest from the WAL,
+// aborting on error.
+// prwe.mu is temporarily acquired for the individual read attempts only,
+// allowing writes to happen while waiting
+func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (*prompb.WriteRequest, error) {
 
-		protoBlob, err = prwe.wal.Read(index)
-		if err == nil { // The read succeeded.
-			req := new(prompb.WriteRequest)
-			if err = proto.Unmarshal(protoBlob, req); err != nil {
-				return nil, err
-			}
+	if prwe == nil {
+		return nil, fmt.Errorf("attempt to read from closed WAL")
+	}
 
-			// Now increment the WAL's read index.
-			prwe.rWALIndex.Add(1)
+	try := func() (*prompb.WriteRequest, error) {
```

Comment:
Could I ask that you move this to be a method as part of the type?

My concern being is that trying to unit test this means a lot of potential set up and keeping things simple and isolated helps checking where things break (or making sure they break correctly).